File: dart/lib/data_structures/linked_list.dart
import 'package:dartz/dartz.dart';
import 'package:equatable/equatable.dart';
import 'package:snowblind/core/types.dart';

class _Node<T> extends Equatable {
  final T value;
  _Node<T> next = null;

  _Node(this.value);

  @override
  List<Object> get props => [value];
}

class LinkedList<T> extends Equatable {
  int _length = 0;
  _Node<T> _head = null;
  _Node<T> _tail = null;

  int get length => _length;

  @override
  List<Object> get props => [_length, _head, _tail];

  Option<T> head() => _head != null ? Some(_head.value) : None();

  Option<T> tail() => _tail != null ? Some(_tail.value) : None();

  LinkedList<T> push(final T value) {
    _length += 1;

    if (_tail == null) {
      _tail = _Node(value);
      _head = _tail;

      return this;
    }

    _tail.next = _Node(value);
    _tail = _tail.next;

    return this;
  }

  void forEach(Consumer<T> consumer) {
    var currentNode = _head;

    while (currentNode != null) {
      consumer(currentNode.value);

      currentNode = currentNode.next;
    }
  }

  U reduce<U>(final U initialValue, final Reducer<T, U> reducer) {
    var result = initialValue;

    var currentNode = _head;

    while (currentNode != null) {
      result = reducer(result, currentNode.value);

      currentNode = currentNode.next;
    }

    return result;
  }

  LinkedList<T> filter(final Predicate<T> predicate) {
    final list = LinkedList<T>();

    var currentNode = _head;

    while (currentNode != null) {
      if (predicate(currentNode.value)) {
        list.push(currentNode.value);
      }

      currentNode = currentNode.next;
    }

    return list;
  }

  LinkedList<U> map<U>(final Morphism<T, U> morphism) {
    final list = LinkedList<U>();

    var currentNode = _head;

    while (currentNode != null) {
      list.push(morphism(currentNode.value));

      currentNode = currentNode.next;
    }

    return list;
  }

  LinkedList<T> concat(final LinkedList<T> other) {
    var list = LinkedList<T>();

    var thisListCurrentNode = _head;

    while (thisListCurrentNode != null) {
      list.push(thisListCurrentNode.value);

      thisListCurrentNode = thisListCurrentNode.next;
    }

    var otherListCurrentNode = other._head;

    while (otherListCurrentNode != null) {
      list.push(otherListCurrentNode.value);

      otherListCurrentNode = otherListCurrentNode.next;
    }

    return list;
  }
}

File: golang/data_structures/lists/LinkedList/LinkedList.go
package linkedlist

import "errors"

type Predicate func(value interface{}) bool
type Mapper func(value interface{}) interface{}
type Reducer func(accumulator interface{}, value interface{}) interface{}
type Consumer func(value interface{})

type Node struct {
	next  *Node
	Value interface{}
}

type LinkedList struct {
	head   *Node
	tail   *Node
	length int
}

func New() *LinkedList {
	var list = &LinkedList{}
	list.length = 0
	return list
}

func (list *LinkedList) Length() int {
	return list.length
}

func (list *LinkedList) IsEmpty() bool {
	return list.Length() == 0
}

func (list *LinkedList) Head() *Node {
	return list.head
}

func (list *LinkedList) Tail() *Node {
	return list.tail
}

func (list *LinkedList) assertIndexIsValid(index int) bool {
	return index >= 0 && index <= list.Length()
}

func (list *LinkedList) At(index int) interface{} {
	if !list.assertIndexIsValid(index) {
		return errors.New("Invalid index")
	}

	var i = 0
	var currentNode = list.Head()

	for i != index {
		i++
		currentNode = currentNode.next
	}

	return currentNode.Value
}

func (list *LinkedList) Push(value interface{}) *LinkedList {
	if list.IsEmpty() {
		list.head = &Node{Value: value}
		list.tail = list.head
	} else {
		list.tail.next = &Node{Value: value}
		list.tail = list.tail.next
	}

	list.length++
	return list
}

func (list *LinkedList) FillRange(start int, end int) *LinkedList {
	for start < end {
		list.Push(start)
		start++
	}

	return list
}

func (list *LinkedList) RemoveOne(predicate Predicate) *LinkedList {
	var previousNode *Node = nil
	var currentNode *Node = list.Head()

	for currentNode != nil {
		if predicate(currentNode.Value) {
			list.length--

			if previousNode == nil {
				list.head = currentNode.next
			} else {
				previousNode.next = currentNode.next
			}

			if currentNode.next == nil {
				list.tail = currentNode
			}

			return list
		}

		previousNode = currentNode
		currentNode = currentNode.next
	}

	return list
}

func (list *LinkedList) RemoveAll(predicate Predicate) *LinkedList {
	var previousNode *Node = nil
	var currentNode *Node = list.Head()

	for currentNode != nil {
		if predicate(currentNode.Value) {
			list.length--

			if previousNode == nil {
				list.head = currentNode.next
			} else {
				previousNode.next = currentNode.next
			}

			if currentNode.next == nil {
				list.tail = currentNode
			}

		}

		previousNode = currentNode
		currentNode = currentNode.next
	}

	return list
}

func (list *LinkedList) Includes(predicate Predicate) bool {
	var currentNode *Node = list.Head()

	for currentNode != nil {
		if predicate(currentNode.Value) {
			return true
		}
		currentNode = currentNode.next
	}

	return false
}

func (list *LinkedList) All(predicate Predicate) bool {
	var currentNode *Node = list.Head()

	for currentNode != nil {
		if !predicate(currentNode.Value) {
			return false
		}
		currentNode = currentNode.next
	}

	return true
}

func (list *LinkedList) Any(predicate Predicate) bool {
	var currentNode *Node = list.Head()

	for currentNode != nil {
		if predicate(currentNode.Value) {
			return true
		}
		currentNode = currentNode.next
	}

	return false
}

func (list *LinkedList) None(predicate Predicate) bool {
	return !list.All(predicate)
}

func (list *LinkedList) Map(mapper Mapper) *LinkedList {
	var newList = &LinkedList{}

	var currentNode *Node = list.Head()

	for currentNode != nil {
		newList.Push(mapper(currentNode.Value))
		currentNode = currentNode.next
	}

	return newList
}

func (list *LinkedList) Filter(predicate Predicate) *LinkedList {
	var newList = &LinkedList{}

	var currentNode *Node = list.Head()

	for currentNode != nil {
		if predicate(currentNode.Value) {
			newList.Push(currentNode.Value)
		}
		currentNode = currentNode.next
	}

	return newList
}

func (list *LinkedList) Reduce(initialValue interface{}, reducer Reducer) interface{} {
	var accumulator interface{} = initialValue
	var currentNode *Node = list.Head()

	for currentNode != nil {
		accumulator = reducer(accumulator, currentNode.Value)
		currentNode = currentNode.next
	}

	return accumulator
}

func (list *LinkedList) ForEach(consumer Consumer) *LinkedList {
	var currentNode *Node = list.Head()

	for currentNode != nil {
		consumer(currentNode.Value)
		currentNode = currentNode.next
	}

	return list
}
